src/Features/LanguageServer/Protocol/Handler/Configuration/DidChangeConfigurationNotificationHandler.cs

```diff
@@ -0,0 +1,212 @@
+            IClientLanguageServerManager clientLanguageServerManager,
+            IAsynchronousOperationListener asynchronousOperationListener)
+        {
+            _lspLogger = logger;
+            _globalOptionService = globalOptionService;
+            _clientLanguageServerManager = clientLanguageServerManager;
+            _registrationId = Guid.NewGuid();
+            _asynchronousOperationListener = asynchronousOperationListener;
+            _configurationItems = GenerateGlobalConfigurationItems();
+        }
+
+        public bool MutatesSolutionState => true;
+
+        public bool RequiresLSPSolution => false;
+
+        public async Task HandleNotificationAsync(DidChangeConfigurationParams request, RequestContext requestContext, CancellationToken cancellationToken)
+        {
+            using var _ = _asynchronousOperationListener.BeginAsyncOperation(Methods.WorkspaceDidChangeConfigurationName);
+            await RefreshOptionsAsync(cancellationToken).ConfigureAwait(false);
+        }
+
+        private async Task RefreshOptionsAsync(CancellationToken cancellationToken)
+        {
+            var configurationsFromClient = await GetConfigurationsAsync(cancellationToken).ConfigureAwait(false);
+            if (configurationsFromClient.IsEmpty)
+            {
+                // Failed to get values from client, do nothing.
+                return;
+            }
+
+            // We always fetch VB and C# value from client if the option is IPerLanguageValuedOption.
+            RoslynDebug.Assert(configurationsFromClient.Length == SupportedOptions.Sum(option => option is IPerLanguageValuedOption ? 2 : 1));
+            var optionsToRefresh = SupportedOptions.SelectManyAsArray(option => option is IPerLanguageValuedOption
+                ? SupportedLanguages.SelectAsArray(language => (option, language))
+                : SpecializedCollections.SingletonEnumerable((option, string.Empty)));
+
+            // LSP ensures the order of result from client should match the order we sent from server.
+            for (var i = 0; i < configurationsFromClient.Length; i++)
+            {
+                var valueFromClient = configurationsFromClient[i];
+                if (valueFromClient == string.Empty)
+                {
+                    // Configuration doesn't exist in client.
+                    continue;
+                }
+
+                var (option, languageName) = optionsToRefresh[i];
+                if (option is IPerLanguageValuedOption perLanguageValuedOption)
+                {
+                    SetOption(option, valueFromClient, languageName);
+                }
+                else
+                {
+                    SetOption(option, valueFromClient);
+                }
+            }
+        }
+
+        private void SetOption(IOption2 option, string valueFromClient, string? languageName = null)
+        {
+            if (option.Definition.Serializer.TryParse(valueFromClient, out var result))
```

Comment:
if we couldn;t parse the option - lets log a warning (since the options are user configured we could get basically any value back so its not an error).